Return error when node info is not found in cache

alculquicondor opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

#79076 migrated from using informer's cache to scheduler cache to increase consistency, but there could be race conditions when a node is removed. Note that we have some checks for nil in earlier calls, but nodes can be removed anytime. This could happen more often when doing fast downscales.

This change returns an API error to mimic the behavior in earlier versions.

**Which issue(s) this PR fixes**:

Fixes #89410

**Special notes for your reviewer**:

This is not a problem in 1.17+, because we use a snapshot of the cache, which remains consistent during the scheduling cycle.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

alculquicondor commented:
/priority critical-urgent

ahg-g commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1308,6 +1307,9 @@ func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod,
 		return false, false, err
 	}
+	if targetPodNode == nil {
+		return false, false, fmt.Errorf("node %s was deleted", targetPod.Spec.NodeName)
```
We should try to be consistent, in line 1386 below a nil node is not considered an error.

ahg-g commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1379,13 +1381,12 @@ func (c *PodAffinityChecker) getMatchingAntiAffinityTopologyPairsOfPods(pod *v1.
 			return nil, err
 		}
+		if existingPodNode == nil {
+			klog.Infof("node %v was deleted", existingPod.Spec.NodeName)
```
this log is unlikely to be useful, and will cause undue suspicion that something is wrong when looking at the scheduler logs.

ahg-g commented on pkg/scheduler/algorithm/priorities/interpod_affinity.go:
```diff
@@ -114,12 +113,12 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *v1.Pod, node
 			return err
 		}
+		if existingPodNode == nil {
+			klog.Infof("node %v was deleted", existingPod.Spec.NodeName)
```
ditto regarding the log message.

ahg-g commented on pkg/scheduler/algorithm/priorities/interpod_affinity.go:
```diff
@@ -114,12 +113,12 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *v1.Pod, node
 	processPod := func(existingPod *v1.Pod) error {
 		existingPodNode, err := ipa.info.GetNodeInfo(existingPod.Spec.NodeName)
 		if err != nil {
```
I would just do this instead of the block below:

```suggestion
		if err != nil || existingPodNode == nil {
```

ahg-g commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1379,13 +1381,12 @@ func (c *PodAffinityChecker) getMatchingAntiAffinityTopologyPairsOfPods(pod *v1.
 	for _, existingPod := range existingPods {
 		existingPodNode, err := c.info.GetNodeInfo(existingPod.Spec.NodeName)
 		if err != nil {
```
doing this is fine and simple because all the pods are from the same node.

```suggestion
		if err != nil || existingPodNode == nil {
```

ahg-g commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1308,6 +1307,9 @@ func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod,
 	if err != nil {
```
```suggestion
	if err != nil || targetPodNode == nil {
```

ahg-g reviewed (commented):
There is another call here: https://github.com/kubernetes/kubernetes/blob/e2e812b5df79e09c2337e4dd21f09cfc967f504d/pkg/scheduler/algorithm/predicates/predicates.go#L1136

alculquicondor committed:
Return error when node info is not found in cache

This can happen when a node deletion happens while a scheduling cycle is ongoing (for example during downscales).
This is the equivalent behavior in 1.15, which used informers caches.

ahg-g commented:
/approve
/lgtm

alculquicondor commented:
/retest

alculquicondor commented:
/retest

alculquicondor commented:
@liggitt do you have any clue what's going on with `pull-kubernetes-e2e-*`? The errors seem totally unrelated to this PR.

liggitt commented:
/retest

alculquicondor commented:
@kubernetes/patch-release-team

This PR is now approved and tests are passing.
